lib/common/schemas.c

```diff
@@ -314,6 +314,57 @@ transform_filter(const struct dirent *entry)
     return 1;
 }
 
+/*!
+ * \internal
+ * \brief Compare transform files based on the version strings in their names
+ *
+ * This is a crude version comparison that relies on the specific structure of
+ * these filenames.
+ *
+ * \retval -1 if \p entry1 sorts before \p entry2
+ * \retval  0 if \p entry1 sorts equal to \p entry2
+ * \retval  1 if \p entry1 sorts after \p entry2
+ *
+ * \note The GNU \c versionsort() function would be perfect here, but it's not
+ *       portable.
+ */
+static int
+compare_transforms(const struct dirent **entry1, const struct dirent **entry2)
+{
+    unsigned char major1 = 0;
+    unsigned char major2 = 0;
+    unsigned char minor1 = 0;
+    unsigned char minor2 = 0;
+    unsigned char order1 = 0;
+    unsigned char order2 = 0;
+
+    // If these made it through the filter, they should be of the right format
+    CRM_LOG_ASSERT(sscanf((*entry1)->d_name, "upgrade-%hhu.%hhu-%hhu.xsl",
+                          &major1, &minor1, &order1) == 3);
+    CRM_LOG_ASSERT(sscanf((*entry2)->d_name, "upgrade-%hhu.%hhu-%hhu.xsl",
+                          &major2, &minor2, &order2) == 3);
+
+    if (major1 < major2) {
+        return -1;
+    } else if (major1 > major2) {
+        return 1;
+    }
+
+    if (minor1 < minor2) {
+        return -1;
+    } else if (minor1 > minor2) {
+        return 1;
+    }
+
+    if (order1 < order2) {
+        return -1;
+    } else if (order1 > order2) {
+        return 1;
+    }
+
+    return 0;
+}
+
```

Comment:
Yes, but:  https://man.freebsd.org/cgi/man.cgi?versionsort(3) which was released in April 2023.  That's pretty recent, but I don't think we make any guarantees about FreeBSD aside from making sure some version builds.  There's no mention of a minimum version in https://projects.clusterlabs.org/T1.

Reply:
Those are "important distributions," but it seems like a bad move to do something that we're almost certain will break recent versions of other distros and Unixes. As far as standards go, `versionsort()` is only part of GNU.

I doubt that this `compare_transforms()` function is as efficient as `versionsort()`, but it does what we need, it's pretty easy to follow, and this isn't something Pacemaker has to do a TON of. We can rely on a very specific structure of version numbers, while `versionsort()` (being a library function) has to be very general.

Reply:
If you're curious, this is what `versionsort()` is doing on the back end.
https://codebrowser.dev/glibc/glibc/string/strverscmp.c.html

Reply:
> Those are "important distributions," but it seems like a bad move to do something that we're almost certain will break recent versions of other distros and Unixes. As far as standards go, `versionsort()` is only part of GNU

I feel like we are creeping up on a rabbit hole of "does pacemaker work if the distribution uses musl/uclibc/..." here.  In general I agree that we shouldn't make things harder on other distributions or OSes, but I also don't want to make things harder on ourselves.  To be thorough, I feel like we'd need to try building with _GNU_SOURCE disabled and see what breaks.

In this specific case, I don't care about using `versionsort()` vs. using our own thing.  We were getting by without it until just recently.

Reply:
> To be thorough, I feel like we'd need to try building with _GNU_SOURCE disabled and see what breaks.

Building with `_GNU_SOURCE` disabled would likely break a lot more than we care about. It doesn't just enable GNU-specific functions. It transitively enables a bunch of other things. Without it, for example, POSIX 2008 functions are unavailable on Linux.

> We were getting by without it until just recently.

Yeah, we had no need for sorting by version until 3.0.0. The schema transform file names were set up differently.

Reply:
I think it's reasonable to try to conform to POSIX as much as possible, to give Pacemaker the best chance of building on as many systems as possible. If this becomes a big problem, I'm open to reconsidering.